### DISCUSSION TITLE
Should a provider invoke another provider side-effect?

### DISCUSSION THREAD

**Discussion by thangmoxielabs** [rrousselGit/riverpod#3523] [Q&A]
```dart
@riverpod
class RandomNumber extends _$RandomNumber {
  @override
  int build() => math.Random();
  void randomize() => state = math.Random();
}

@riverpod
Future<double> keepRandomUntil(KeepRandomUntilRef ref, double endValue) async {
  final currentValue = ref.watch(randomNumberProvider);
  if (currentValue != endValue) {
    ref.read(randomNumberProvider.notifier).randomize();
  }
}
```
Is this pattern acceptable, for a provider to trigger another provider which could trigger that provider again until a certain condition is met?
P/s: I haven't tried this on an editor, so excuse me for syntax errors

**Comment by rrousselGit** [comment]
No. It falls under: https://riverpod.dev/docs/essentials/do_dont#dont-perform-side-effects-during-the-initialization-of-a-provider

**Comment by thangmoxielabs** [comment]
Yea that makes sense. Could you recommend a way to achieve the same effect?
I was going to refactor the `keepRandomUntil` into a class and put everything into another function, but I guess that violates https://riverpod.dev/docs/essentials/combining_requests#the-refwatch-method
```dart
@riverpod
class KeepRandomUntil extends _$KeepRandomUntil {
  @override
  void build() {};
  void execute(double endValue) {
    final currentValue = ref.watch(randomNumberProvider); // BAD, no watch in non-build function
    if (currentValue != endValue) {
      ref.read(randomNumberProvider.notifier).randomize();
    }
  }
}
```

**Reply by rrousselGit** [reply]
Your logic is reversed. You probably should have randomNumberProvider watch keepRandomUntil
